File: examples/bun/demo_cancellation.ts
/**
 * mp4decrypt-api - Asynchronous Session Cancellation Demo (Bun)
 * ============================================================
 *
 * Demonstrates interrupting a decryption pipeline mid-flight
 * via session.cancel(). The operation safely unwinds resources
 * and returns cancellation error code -2.
 *
 * Run with:
 *   bun examples/bun/demo_cancellation.ts
 */

import { Mp4DecryptSession, Mp4LogLevel } from "../../sdk/bun/mp4decrypt_sdk";

async function main() {
  console.log("===================================================================");
  console.log("    MP4DECRYPT-API - ASYNCHRONOUS CANCELLATION DEMO (BUN)          ");
  console.log("===================================================================");

  const baseUrl = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey";
  const kidHex = "9eb4050de44b4802932e27d75083e266";
  const keyHex = "166634c675823c235a4a9446fad52e4d";

  console.log("Fetching encrypted media chunks...");
  const [initResp, segResp] = await Promise.all([
    fetch(`${baseUrl}/15/init.mp4`),
    fetch(`${baseUrl}/15/0001.m4s`),
  ]);
  const initBytes = new Uint8Array(await initResp.arrayBuffer());
  const segBytes = new Uint8Array(await segResp.arrayBuffer());
  const inputBuffer = new Uint8Array(initBytes.byteLength + segBytes.byteLength);
  inputBuffer.set(initBytes, 0);
  inputBuffer.set(segBytes, initBytes.byteLength);

  const session = new Mp4DecryptSession();
  try {
    session.addKey(kidHex, keyHex);
    session.setLogLevel(Mp4LogLevel.DEBUG);

    session.onLog((lvl, msg) => {
      console.log(`  [Native Log] ${msg}`);
    });

    let cancellationTriggered = false;
    let inPos = 0;

    console.log("\nStarting streaming decryption with planned mid-flight cancellation...");
    try {
      session.decryptCustom({
        input: {
          read: (buf, count) => {
            const available = Math.min(count, inputBuffer.length - inPos);
            if (available <= 0) return 0;
            buf.set(inputBuffer.subarray(inPos, inPos + available));
            inPos += available;

            // Trigger cancellation after reading the first chunk
            if (!cancellationTriggered && inPos >= 16) {
              console.log(`  [Custom Stream] Read ${inPos} bytes -> triggering session.cancel()!`);
              cancellationTriggered = true;
              session.cancel();
            }
            return available;
          },
          seek: (offset) => {
            inPos = offset;
            return 0;
          },
          tell: () => inPos,
          size: () => inputBuffer.length,
        },
        output: {
          write: (_buf, count) => count,
          seek: () => 0,
          tell: () => 0,
          size: () => 0,
        },
      });

      console.error("ERROR: Decryption was expected to be cancelled but finished successfully!");
    } catch (err: any) {
      console.log(`\nExpected cancellation caught: ${err.message}`);
      if (err.message.includes("cancelled") || cancellationTriggered) {
        console.log("\nSUCCESS: Session cancellation operated as intended!");
      }
    }
  } finally {
    session.destroy();
  }
}

main().catch(console.error);

File: examples/python/demo_cancellation.py
#!/usr/bin/env python3
"""
mp4decrypt-api - Asynchronous Session Cancellation Demo (Python)
================================================================

Demonstrates interrupting a decryption operation mid-flight using
session.cancel() from a stream callback or external thread.

Run with:
    python examples/python/demo_cancellation.py
"""

import io
import os
import sys
import urllib.request

# Ensure sdk/python is importable
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..")))
from sdk.python.mp4decrypt_sdk import Mp4DecryptSession, Mp4LogLevel


class CancellingInputStream(io.BytesIO):
    def __init__(self, data: bytes, session: Mp4DecryptSession):
        super().__init__(data)
        self.session = session
        self.cancelled = False

    def read(self, size=-1):
        # Once some bytes have been read, trigger cancellation
        pos = self.tell()
        if pos >= 16 and not self.cancelled:
            print(f"  [Custom Stream] Read {pos} bytes -> triggering session.cancel()!")
            self.cancelled = True
            self.session.cancel()
        return super().read(size)


def main():
    print("=" * 67)
    print("    MP4DECRYPT-API - ASYNCHRONOUS CANCELLATION (PYTHON)            ")
    print("=" * 67)

    base_url = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey"
    kid_hex = "9eb4050de44b4802932e27d75083e266"
    key_hex = "166634c675823c235a4a9446fad52e4d"

    print("Fetching encrypted media chunks...")
    init_bytes = urllib.request.urlopen(f"{base_url}/15/init.mp4").read()
    seg_bytes = urllib.request.urlopen(f"{base_url}/15/0001.m4s").read()
    test_buffer = init_bytes + seg_bytes

    print("\nStarting streaming decryption with planned mid-flight cancellation...")
    with Mp4DecryptSession() as session:
        session.add_key(kid_hex, key_hex)
        session.set_log_level(Mp4LogLevel.DEBUG)
        session.on_log(lambda lvl, msg: print(f"  [Native Log] {msg}"))

        in_stream = CancellingInputStream(test_buffer, session)
        out_stream = io.BytesIO()

        try:
            session.decrypt_custom(in_stream, out_stream)
            print("ERROR: Decryption was expected to fail with cancellation error!")
        except RuntimeError as e:
            print(f"\nExpected cancellation caught: {e}")
            if "cancelled" in str(e).lower() or in_stream.cancelled:
                print("\nSUCCESS: Session cancellation operated as intended!")


if __name__ == "__main__":
    main()
